### PR TITLE
feat: lsp "go to definition" for modules

### LINKED ISSUE #5405: Let LSP "go to definition" work in path segments
### Problem

"Go to definition" currently doesn't work for path segments. For example if you have `use foo::bar` and click on `foo`, it doesn't take you to the `foo` module. Similarly, if you have a call like `foo::bar()`, clicking on `foo` doesn't do anything. Then, if you have `mod foo;`, clicking on `foo` doesn't do anything either.

### Happy Case

Clicking on any part of a path, be it in a `mod` statement, `use` statement, a call or a reference to a type, etc., should take you to the module declaration. This should work both for modules declared in a file as well as inline modules (`mod foo { ... }`).

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #5405

## Summary

Now clicking on module segments works as expected.

https://github.com/noir-lang/noir/assets/209371/e8968eea-ad0b-4d85-9500-db31e46009a0

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs [hunk 1]
```diff
@@ -330,7 +330,30 @@ impl DefCollector {
...
+                    &collected_import,
+                    &context.def_maps,
+                    &mut Some(&mut references),
+                );
+
+                let current_def_map = context.def_maps.get(&crate_id).unwrap();
+                let file_id = current_def_map.file_id(module_id);
+
+                for (referenced, ident) in
+                    references.iter().zip(collected_import.clone().path.segments)
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs:346]:
```suggestion
                for (referenced, ident) in references.iter().zip(&collected_import.path.segments)
```

We can avoid a clone here we we only need a reference to the path segments to get their spans.

File: compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs [hunk 2]
```diff
@@ -643,13 +672,14 @@ impl<'a> ModCollector<'a> {
...
         mod_name: &Ident,
-        file_id: FileId,
+        mod_location: Location,
         add_to_parent_scope: bool,
         is_contract: bool,
-    ) -> Result<LocalModuleId, DefCollectorErrorKind> {
+    ) -> Result<ModuleId, DefCollectorErrorKind> {
         let parent = Some(self.module_id);
-        let location = Location::new(mod_name.span(), file_id);
+        let location = Location::new(mod_name.span(), mod_location.file);
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs:682]: It looks like we have a lot of overlap between `mod_name`, `mod_location` and `location`. Do we really need all three of these? Seems like we can remove one or the other of `mod_location` and `location` at least. (maybe so for structs/submodules but not modules in a separate file?)

**Review by TomAFrench — COMMENTED** [review]
Looks good, I would like to check that all the location info we're passing around is necessary and not redundant however.

File: compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs (hunk 2 above)
Comment (TomAFrench) [reply compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs:682]: > I don't know what that part of contracts is used for.

This is use to display codelenses above the `contract Foo` line so we're going to want to maintain this.

I'm inclined to just make a note of this so we can merge these PRs and address this as potential tech debt in the future.

**Review by TomAFrench — APPROVED** [review]

File: compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs (hunk 2 above)
Comment (asterite) [reply compiler/noirc_frontend/src/hir/def_collector/dc_mod.rs:682]: I'll add a note 👍

**Pull request merged**